toolchain/driver/runtimes_cache_test.cpp

```diff
@@ -0,0 +1,707 @@
+  // directory and target to simulate two versions or builds of the Carbon
+  // toolchain.
+  auto custom_install_dir =
+      *tmp_dir_.CreateDirectories("custom_install/lib/carbon");
+  custom_install_dir.WriteFileFromString("carbon_install.txt", "diff digest")
+      .Check();
+  custom_install_dir.WriteFileFromString("install_digest.txt", "abcd").Check();
+  InstallPaths install2 =
+      InstallPaths::Make((tmp_dir_.abs_path() / "custom_install").native());
+  auto cache2 = *Runtimes::Cache::MakeCustom(install2, tmp_dir_.abs_path());
+  auto runtimes2 = *cache2.Lookup({.target = target});
+
+  // The parent paths of these runtimes should be the same.
+  EXPECT_THAT(runtimes1.base_path().parent_path(),
+              Eq(runtimes2.base_path().parent_path()));
+
+  // But the base paths for these two runtimes should differ due to cache key
+  // differences.
+  EXPECT_THAT(runtimes1.base_path(), Ne(runtimes2.base_path()));
+}
+
+TEST_F(RuntimesCacheTest, ConcurrentBuilds) {
+  const std::string target = "aarch64-unknown-unknown";
+  auto runtimes1 = *cache_.Lookup({.target = target});
+
+  // Build a second cache and runtimes pointing at the same directory and target
+  // to simulate concurrent processes.
+  auto cache2 = *Runtimes::Cache::MakeCustom(install_, tmp_dir_.abs_path());
+  auto runtimes2 = *cache2.Lookup({.target = target});
+
+  // Start the first build, this will lock the directory.
+  auto build_result1 = runtimes1.Build(Runtimes::ClangResourceDir);
+  ASSERT_THAT(build_result1, IsSuccess(VariantWith<Runtimes::Builder>(_)));
+  auto builder1 = std::get<Runtimes::Builder>(*std::move(build_result1));
+  EXPECT_THAT(builder1.dir().WriteFileFromString("runtime_file", "build1"),
+              IsSuccess(_));
+
+  // Start the second build in a separate thread so that it can block while we
+  // finish the first build. The only result we'll need at the end is the built
+  // path.
+  std::filesystem::path build2_path;
+  auto build2_lambda = [&build2_path, &runtimes2] {
+    // Typically building here will try to acquire the same file lock acquired
+    // with the first build. However, the file locking is always _advisory_ and
+    // may fail. As a consequence we can't make assumptions about whether this
+    // blocks or not.
+    auto build_result2 = runtimes2.Build(Runtimes::ClangResourceDir);
+    ASSERT_THAT(build_result2, IsSuccess(_));
+    if (std::holds_alternative<std::filesystem::path>(*build_result2)) {
+      // In the common case, we blocked on a file lock and find the first built
+      // result directly. Save it.
+      build2_path = std::get<std::filesystem::path>(*std::move(build_result2));
+    } else {
+      // In rare cases, the initial build will fail to acquire the file lock.
+      // The entire build process is designed specifically to be resilient to
+      // that so we should still succeed, but now we need to handle building in
+      // this thread as well.
```

Comment:
Optional: I'm inclined to be slightly more explicit about the fact that this is an exception to the usual rule that flaky test failures are false positives.

```suggestion
      // this thread as well. Note that a true failure here may only
      // show up intermittently.
```

Reply:
I like that, thanks.